installplan_operator/main.py

```diff
@@ -0,0 +1,229 @@
+
+
+class FileWatcher(threading.Thread):
+    '''Watch for filesystem events in config_dir'''
+
+    def __init__(self, q):
+        self.q = q
+        super().__init__(daemon=True)
+
+    def run(self):
+        for event in watch(config.config_dir):
+            self.q.put(('fs', event))
+
+
+class Approver:
+    def __init__(self):
+        self.oc = create_openshift_client()
+        self.q = queue.Queue()
+
+    def start_watchers(self):
+        self.tasks = [
+            FileWatcher(self.q),
+            SubscriptionWatcher(self.q, self),
+        ]
+
+        for task in self.tasks:
+            task.start()
+
+    @cached_property
+    def subscriptions(self):
+        '''This is an API endpoint for listing/getting/watching subscriptions'''
+
+        return self.oc.resources.get(
+            api_version='operators.coreos.com/v1alpha1',
+            kind='Subscription'
+        )
+
+    @cached_property
+    def installplans(self):
+        '''This is an API endpoint for listing/getting/watching installplans'''
+
+        return self.oc.resources.get(
+            api_version='operators.coreos.com/v1alpha1',
+            kind='InstallPlan'
+        )
+
+    def events(self):
+        '''An iterator that yields events from the queue.
+
+        This method will synthesize a timeout event if no events
+        are received from the queue after max_interval seconds.
+        '''
+
+        while True:
+            # Wait up to max_interval seconds for an event
+            try:
+                event = self.q.get(timeout=config.max_interval)
+            except queue.Empty:
+                event = ('timeout',)
+
+            yield event
+
+    def loop(self):
+        self.start_watchers()
+
+        t_last = 0
+        for event in self.events():
+            t_start = time.time()
+            t_delta = t_start - t_last
+
+            # Ignore triggers if we last checked less than min_interval
+            # seconds ago.
+            if t_delta < config.min_interval:
+                LOG.debug('ignoring %s event (too soon), t_last=%s, t_start=%s, t_delta=%s',
+                          event[0],
+                          t_last,
+                          t_start,
+                          t_delta)
+                continue
+
+            LOG.info('subscription check triggered by %s event', event[0])
+            sub_count = self.process_update_specs()
+
+            t_end = time.time()
+            t_last = t_start
+            t_delta = t_end - t_start
+
+            LOG.info('finished subscription check; %s subscriptions in %s seconds',
+                     sub_count,
+                     t_delta)
+
+    def process_update_specs(self):
+        '''Iterate through update specifications in config_dir'''
+
+        count = 0
+
+        for specfile in config.config_dir.iterdir():
+            ext = specfile.name.rsplit('.', 1)[-1]
+            if specfile.is_file() and ext in ['yaml', 'yml']:
+                count += 1
```

Comment:
Wouldn't be better to use `glob.glob` here instead?

Reply:
I don't think there's a big difference either way. I'm happy to update it to use `glob.glob` instead.

Reply:
I was thinking if we can simplify this whole function into more pythonic way. Since `config.config_dir` is already `Path`, we can do something like this:

```python
def process_update_specs(self):
    specfiles = list(config.config_dir.glob('*.yaml')) + list(config.config_dir.glob('*.yml'))

    for specfile in specfiles:
        try:
            self.process_update_spec(specfile)
        except openshift_exc.NotFoundError as err:
            LOG.error('unable to find requested resource: %s', err.summary())

    return len(specfiles)
```

If you want you can also `chain` those generators to make it even more pythonic.

I think this is more resilient, since it can't fail you in cases like when the file is named `yaml` the `ext = specfile.name.rsplit('.', 1)[-1]` would return `yaml`, but this is not a valid input...

Reply:
I've updated the code to use pathlib's `glob` method, and added a test to make sure the file discovery is working as expected.